server/src/main/java/org/elasticsearch/cluster/coordination/MasterHistoryService.java

```diff
@@ -0,0 +1,119 @@
+     */
+    public MasterHistory getLocalMasterHistory() {
+        return localMasterHistory;
+    }
+
+    /**
+     * This method returns a static view of the MasterHistory on the node given. This MasterHistory is static in that it will not be
+     * updated even if the ClusterState is updated on this node or the remote node. The history is retrieved asynchronously, and only if
+     * requestRemoteMasterHistory has been called for this node. If anything has gone wrong fetching it, or if it has not been fetched in
+     * time for the current request, the returned value will be null.
+     * @param node The node whose MasterHistory we want
+     * @return The MasterHistory from the remote node's point of view. This MasterHistory object will not be updated with future changes
+     */
+    @Nullable
+    public List<DiscoveryNode> getRemoteMasterHistory(DiscoveryNode node) {
+        return nodeToHistoryMap.get(node);
+    }
+
+    /**
+     * This method attempts to fetch the master history from the requested node. If we are able to successfully fetch it, it will be
+     * available in a later call to getRemoteMasterHistory. The client is not notified if or when the remote history is successfully
+     * retrieved. This method only fetches the remote master history once, and it is never updated unless this method is called again.
+     * This is a remote call, so clients should avoid calling it any more often than necessary.
+     * @param node The node whose view of the master history we want to fetch
+     */
+    public void requestRemoteMasterHistory(DiscoveryNode node) {
+        long startTime = System.nanoTime();
+        transportService.openConnection(
+            // Note: This connection is explicitly closed onResponse or onFailure of the request below
+            node,
+            ConnectionProfile.buildDefaultConnectionProfile(clusterService.getSettings()),
+            new ActionListener<>() {
+                @Override
+                public void onResponse(Transport.Connection connection) {
+                    logger.trace("Opened connection to {}, making master history request", node);
+                    transportService.sendRequest(
+                        node,
+                        MasterHistoryAction.NAME,
+                        new MasterHistoryAction.Request(),
+                        new ActionListenerResponseHandler<>(new ActionListener<>() {
+
+                            @Override
+                            public void onResponse(MasterHistoryAction.Response response) {
+                                connection.close();
+                                long endTime = System.nanoTime();
+                                logger.trace("Received history from {} in {}", node, TimeValue.timeValueNanos(endTime - startTime));
+                                nodeToHistoryMap.put(node, response.getMasterHistory());
```

Comment:
We may send two requests concurrently to the same node and get the answers back in the other order, with the stale response overwriting the fresher one. Not sure if this matters. If it doesn't, we should at least comment that this behaviour is deliberate.

Reply:
That would happen if we have flapped null 4 at least times locally (so call 1 would be made) and then we went not-null and then null again. If that happens I think we don't really care what we get back from the remote master because it'll probably have the same shape either way (either it similarly shows lots of flapping in both responses, or it shows that everything is ok in both responses). Right?

Reply:
Assuming it's fine, I added a note to the javadoc.